### PR TITLE
[11.0][MIG] kpi

### PR DESCRIPTION
**Pull request by AdriaGForgeFlow**
Migration from version 9.0 to 11.0

### CODE REVIEW

File: oca_dependencies.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+server-backend
```
Comment (JordiBForgeFlow) [inline oca_dependencies.txt]: Runbot fails. Instead of this line, try 
`server-backend https://github.com/OCA/server-backend`

Comment (AdriaGForgeFlow) [reply oca_dependencies.txt]: Done. Thanks :+1:

**Comment by JordiBForgeFlow** [comment]
@moylop260 @pedrobaeza not sure why runbot fails on this PR.. perhap you have found this issue before?

**Comment by pedrobaeza** [comment]
It seems a glitch. I have restarted the build to see.

**Comment by pedrobaeza** [comment]
Well, same problem... so no clue. @moylop260 your turn, hehe

File: kpi/README.rst [hunk 2]
```diff
@@ -0,0 +1,78 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
```
Comment (etobella) [inline kpi/README.rst]: Can you divide the readme following the new rules?
https://github.com/OCA/maintainer-tools/tree/master/template/module/readme

**Comment by etobella** [comment]
It is interesting, but you could change the view from tree to kanban. This way it is easier to view it and you can make it prettier (also, you can add much more options).
If you want a pretty option in order to view the KPI, you can use http://dashing.io. It is interesting to have the last update in order to be sure when the KPI was computed.

**Comment by pedrobaeza** [comment]
There's already a similar dashboard widget in https://github.com/OCA/web/tree/8.0/web_dashboard_tile, although I think it's enough with kanban with proper styling.

**Comment by JordiBForgeFlow** [comment]
@ageficent can you replace the images in kpi/images with ones from v11? The ones shown now are from a prehistoric version :)

**Comment by etobella** [comment]
I have checked the runbot and the problem seems related to the inclusion of server-backend dependency.

I checked that mysqlclient-dev is being installed  on server-backend and it is necessary on the runbot in order to install all python dependencies (mysqlclient). However, mysqlclient-dev is not installed on server-tools runbot.  I don't know how to add this dependency, but I think that it will solve the problem.
Maybe it is enough to add the information to .travis.yml (https://github.com/OCA/server-backend/blob/11.0/.travis.yml#L15)
@ageficent @jbeficent

**Comment by etobella** [comment]
Missing dependency to web

**Comment by JordiBForgeFlow** [comment]
@etobella fixed!

File: kpi/views/kpi_views.xml [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                </sheet>
+            </form>
+        </field>
+    </record>
+
+    <record model="ir.actions.act_window" id="open_kpi_dashboard">
+        <field name="name">KPI Dashboard</field>
+        <field name="res_model">kpi</field>
+        <field name="view_type">form</field>
+        <field name="view_mode">kanban,form</field>
```
Comment (etobella) [inline kpi/views/kpi_views.xml:133]: Just kanban, if you add form, you are may enter to the configuration, but the usual user only needs to access to the kanban

**Review by etobella — COMMENTED** [review]
LGTM 👍 Just a small comment.

**Comment by JordiBForgeFlow** [comment]
@etobella we discussed about this. But then the user cannot see the history of past kpi values. Which is interesting to a user.

You will notice that as a user without permissions you cannot create/edit/delete anything, and you cannot see the sql code. Just the history and the thresholds.

**Comment by etobella** [comment]
I see, then it's fine for me 😃

**Review by etobella — APPROVED** [review]

**Comment by JordiBForgeFlow** [comment]
@dreispt You are listed as author of this module. Are you good with the improvements described?

File: kpi/security/ir.model.access.csv [hunk 4]
```diff
@@ -4,8 +4,8 @@
 "access_kpi_category_user","kpi.category.user","model_kpi_category","base.group_user",1,0,0,0
 "access_kpi_threshold_user","kpi.threshold.user","model_kpi_threshold","base.group_user",1,0,0,0
 "access_kpi_threshold_range_user","kpi.threshold.range.user","model_kpi_threshold_range","base.group_user",1,0,0,0
-"access_kpi_manager","kpi.manager","model_kpi","base.group_user",1,1,1,1
-"access_kpi_category_manager","kpi.category.manager","model_kpi_category","base.group_user",1,1,1,1
-"access_kpi_threshold_manager","kpi.threshold.manager","model_kpi_threshold","base.group_user",1,1,1,1
-"access_kpi_threshold_range_manager","kpi.threshold.range.manager","model_kpi_threshold_range","base.group_user",1,1,1,1
+"access_kpi_manager","kpi.manager","model_kpi","kpi.group_kpi_manager",1,1,1,1
+"access_kpi_category_manager","kpi.category.manager","model_kpi_category","kpi.group_kpi_manager",1,1,1,1
+"access_kpi_threshold_manager","kpi.threshold.manager","model_kpi_threshold","kpi.group_kpi_manager",1,1,1,1
```
Comment (AdriaGForgeFlow) [inline kpi/security/ir.model.access.csv:9]: I think there's a missing line for access_kpi_history_manager:
"access_kpi_history_manager","kpi.history.user","model_kpi_history","kpi.group_kpi_manager",1,1,1,1

**Review by AdriaGForgeFlow — COMMENTED** [review]
Observation

File: kpi/security/ir.model.access.csv (hunk 4 above)
Comment (etobella) [reply kpi/security/ir.model.access.csv:9]: Sure!, but maybe the user should be able to create also the history if he can refresh the data.
Just a question, all users should be able to access the KPIs? Or a group should be created?

File: kpi/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+# Copyright 2012 - Now Savoir-faire Linux <https://www.savoirfairelinux.com/>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Key Performance Indicator",
+    "version": "11.0.1.1.0",
```
Comment (etobella) [inline kpi/__manifest__.py]: That should be 11.0.1.0.0

File: kpi/models/kpi.py [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+                obj.value = 0
+                obj.color = '#FFFFFF'
+                obj.last_execution = False
+
+    @api.multi
+    def compute_kpi_value(self):
+        for obj in self:
+            kpi_value = 0
+            if obj.kpi_code:
+                if obj.kpi_type == 'local' and is_sql_or_ddl_statement(
```
Comment (etobella) [inline kpi/models/kpi.py]: Maybe, it should be better to add a hook if someone wants to make another system of Kpis.

Comment (etobella) [reply kpi/models/kpi.py]: The function should return the dict with value, color and kpi_id.

This way, we could use the python code to compute the color dynamically (compare with Previous year for example)

File: kpi/models/kpi.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        for obj in self:
+            history_vals = obj._get_kpi_value()
+            history_obj = self.env['kpi.history']
+            history_obj.create(history_vals)
+        return True
+
+    @api.multi
+    def update_next_execution_date(self):
+        for obj in self:
+            if obj.periodicity_uom == 'hour':
```
Comment (etobella) [inline kpi/models/kpi.py]: It makes sense to add the option of minutes periodicity?

Comment (JordiBForgeFlow) [reply kpi/models/kpi.py]: Now added!

File: kpi/models/kpi.py (hunk 6 above)
Comment (JordiBForgeFlow) [reply kpi/models/kpi.py]: @etobella Done!

**Comment by JordiBForgeFlow** [comment]
@ageficent Can you review the latest commits and also port them to the v12 PR?

File: kpi/models/kpi.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+            }
+        res.update({'kpi_id': self.id})
+        return res
+
+    @api.multi
+    def compute_kpi_value(self):
+        for obj in self:
+            history_vals = obj._get_kpi_value()
+            history_obj = self.env['kpi.history']
+            history_obj.sudo().create(history_vals)
```
Comment (etobella) [inline kpi/models/kpi.py:189]: Maybe should be better to give permissions to the KPI Manager user. Isn't it?

**Comment by pedrobaeza** [comment]
You can move this the same to OCA/reporting-engine.

**Comment by AdriaGForgeFlow** [comment]
> @ageficent Can you review the latest commits and also port them to the v12 PR?

Changes LGTM :+1: and have been ported to v12 here: [OCA/reporting-engine#267](https://github.com/OCA/reporting-engine/pull/267)

File: kpi/models/kpi.py [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+            obj.next_execution_date = new_date.strftime(DATETIME_FORMAT)
+
+        return True
+
+    # Method called by the scheduler
+    @api.model
+    def update_kpi_value(self):
+        filters = [
+            '&',
```
Comment (etobella) [inline kpi/models/kpi.py]: The query is not correct. It should be:
[
 ('active', '=', True),
'|',
('next_execution_date', '<=', datetime.now().strftime(DATETIME_FORMAT)),
('next_execution_date', '=', False),
]

**Review by etobella — CHANGES_REQUESTED** [review]
The cron update is not working

File: kpi/models/kpi.py (hunk 9 above)
Comment (JordiBForgeFlow) [reply kpi/models/kpi.py]: This is now fixed!

**Comment by JordiBForgeFlow** [comment]
@etobella in the last commit we introduce a small migration to new values for select field 'periodicity_uom' so that the values are in line with ir.cron 'interval_type'.

We also introduce a separate cron that will update the 'Update KPI values' frequency if there exist KPIs with a different max frequency.

So, if you have a KPI that is to be run every minute, the cron to update kpis will be updated to be executed every minute for all kpi's.

**Comment by JordiBForgeFlow** [comment]
We found that the option to create a cron that updates the first cron is not good, and we'll revert it.

**Review by etobella — APPROVED** [review]
LGTM :+1:

**Pull request closed without merging**